File: alumnium/tools/base_tool.py
from pydantic import BaseModel

from alumnium.drivers.base_driver import BaseDriver


class BaseTool(BaseModel):
    @classmethod
    def execute_tool_call(
        cls,
        tool_call: dict,
        tools: list["BaseTool"],
        driver: BaseDriver,
    ):
        tool = tools[tool_call["name"]](**tool_call["args"])
        tool.invoke(driver)

File: packages/typescript/src/tools/BaseTool.ts
import { BaseDriver } from "../drivers/BaseDriver.js";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ToolClass = new (...args: any[]) => BaseTool;

export interface ToolCall {
  name: string;
  args: Record<string, unknown>;
}

export abstract class BaseTool {
  abstract invoke(driver: BaseDriver): void | Promise<void>;

  static async executeToolCall(
    toolCall: ToolCall,
    tools: Record<string, ToolClass>,
    driver: BaseDriver
  ): Promise<void> {
    const ToolClass = tools[toolCall.name];
    if (!ToolClass) {
      throw new Error(`Tool ${toolCall.name} not found`);
    }

    const tool = new ToolClass(toolCall.args);
    await tool.invoke(driver);
  }
}
